src/frontend/flight/sync/writer.rs

```diff
@@ -193,15 +183,20 @@ impl SeafowlDataSyncManager {
             })
             .or_insert(IndexMap::from([(sequence_number, sequence)]));
 
-        // Finally upsert the new sync item for this location
-        let size = batches
-            .iter()
-            .fold(0, |bytes, batch| bytes + batch.get_array_memory_size());
+        let batch = compact_batches(&sync_schema, batches)?;
```

Comment:
Maybe we should do this compaction once on flush instead of on every write. Otherwise in a chain of writes we'd be loading the (constantly growing) staged in-memory batch into a hashmap, then transposing back into rows on every write? Although in that case we won't support running queries with unflushed in-memory data.

Reply:
No no, it doesn't work like that. The record batch(es) for each sync are compacted and stored individually as `DataSyncItem`s in a `DataSyncCollection` vector, meaning the previously compacted batches are never touched again until flushing. So we don't do any more work than if we do it lazily (since we'd need to do it then anyway).

The idea was to compress the size eagerly, so that we keep as low memory profile as possible.

Reply:
Ah I see. What happens if there's a PK changing UPDATE in a future batch that affects a batch that we already compacted?

Reply:
Any PK changing UPDATEs in a future batch must reference the last new PK from a preceding batch (which is the one we keep), so compaction shouldn't affect it. This should be guaranteed by source database PK conflict detection mechanism.

In effect the problem is self-contained, i.e. independent of any upcoming changes.